### PR TITLE
Enhancement: Add International Fixed chronology

### PR DESCRIPTION
**Pull request by catull**
Implemented IFC - chronology for International Fixed calendar, described in http://en.wikipedia.org/wiki/International_Fixed_Calendar.
It is a bit similar to Pax, has the same leap year rule as the Gregorian calendar.
However, it consists of 13 months with each 28 days; leap day is inserted between end of month 6 and beginning of month 7.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,15 +3,17 @@
...
 /dist/
 /test-output/
-/temp*
 /.settings/
 /nbproject/private/
 /target/
 .classpath
 .project
 .checkstyle
-*.class
```
Comment (Clockwork-Muse) [inline .gitignore]: I realize `.class` files are only built in the `target` directory, but we probably don't want to make these deletions anyways.  Changes to the `.gitignore` file should probably be part of a separate PR, as they are unrelated to these feature changes.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+ * <li>day-of-year - There are 365 days in a standard International Fixed year and 366 days in a leap year.
+ *  The days are numbered from 1 to 365 / 366 accordingly.
+ * <li>leap-year - Leap years occur every 4 years, but skips 3 out of four centuries, i.e. when the century is not divisible by 400.
+ *  This is the same rule in use for the Gregorian calendar.
+ * </ul>
+ *
+ * <h3>Implementation Requirements</h3>
+ * This class is immutable and thread-safe.
+ */
+public class InternationalFixedChronology extends AbstractChronology implements Serializable {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: Class is not immutable if you don't declare it `final`

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    /**
+     * Singleton instance for the International fixed chronology.
+     */
+    public static final InternationalFixedChronology INSTANCE = new InternationalFixedChronology ();
+
+    /**
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: The existing implementations aren't putting spaces between member variable declaration and the comments of the next one.

Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: ...Okay, _some_ of them aren't.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+     * There are five 400 year cycles from year zero to 2000.
+     * There are 7 leap years from 1970 to 2000.
+     */
+    static final long DAYS_0000_TO_1970 = (DAYS_PER_CYCLE * 5L) - (30L * 365L + 7L);
+
+    /**
+     * Private constructor, that is public to satisfy the {@code ServiceLoader}.
+     * Use the singleton {@link #INSTANCE} instead.
+     */
+    public InternationalFixedChronology () {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: Constructor should be annotated `@Deprecated`, and commented to match.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+     * Each chronology must refer to an enum or similar singleton to provide the era values.
+     * <p>
+     * This method returns the singleton era of the correct type for the specified era value.
+     *
+     * @param eraValue the era value
+     * @return the calendar system era, not null
+     * @throws DateTimeException if unable to create the era
+     */
+    @Override
+    public Era eraOf (final int eraValue) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: Since you're overriding the era, please also change the return type.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+    @Override
+    public String getId () {
+        return "Ifc";
+    }
+
+    /**
+     * Gets the calendar type of the underlying calendar system - 'ifc'.
+     * <p>
+     * The <em>Unicode Locale Data Markup Language (LDML)</em> specification
+     * does not define an identifier for the Pax calendar, but were it to
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: This is not the Pax calendar.  (It's nice to be used as a base, though).

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+     *                            such as if the year is invalid for the era
+     * @throws ClassCastException if the {@code era} is not of the correct type for the chronology
+     */
+    @Override
+    public int prolepticYear (final Era era, final int yearOfEra) {
+        if (!(era instanceof InternationalFixedEra)) {
+            throw new ClassCastException ("Era must be InternationalFixedEra");
+        }
+
+        if (1 > yearOfEra) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: I get where you're going here, but it's still backwards.  Further, the exception message doesn't really match the condition you check (0 is positive for integers).  I also recommend not putting any special emphasis on `"must"`.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+     */
+    @SuppressWarnings ("static-method")
+    private Object readResolve () {
+        return INSTANCE;
+    }
+
+    //-----------------------------------------------------------------------
+
+    /**
+     * Obtains a local date in Julian calendar system from the
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: This is not the Julian calendar system either.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+     * @param yearOfEra  the year-of-era
+     * @param month      the month-of-year
+     * @param dayOfMonth the day-of-month
+     * @return the Julian local date, not null
+     * @throws DateTimeException if unable to create the date
+     * @throws ClassCastException if the {@code era} is not a {@code JulianEra}
+     */
+    @Override
+    public InternationalFixedDate date (final Era era, final int yearOfEra, final int month, final int dayOfMonth) {
+        return date (prolepticYear (era, yearOfEra), month, dayOfMonth);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: The current style/formatting doesn't put a space between the method name and the opening parenthesis.  Checkstyle should be catching this.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+ * Moses B. Cotsworth, who presented it in 1902.</p>
+ * <p>
+ * It provides for a year of 13 months of 28 days each, with one or two days a year belonging to no month or week.
+ * It is therefore a perennial calendar, with every date fixed always on the same weekday.
+ * Though it was never officially adopted in any country, it was the official calendar of the Eastman Kodak Company
+ * from 1928 to 1989.</p>
+ * <p>
+ * This date operates using the {@linkplain InternationalFixedChronology International fixed calendar}.
+ * This calendar system is a proposed reform calendar system, and is not in common use.
+ * The International fixed differs from the Gregorian in terms of month count and length, and the leap year rule.
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: "The International" sounds a little off (maybe pretentious?).  It does not differ from the Gregorian calendar in terms of the leap-year rule.  Given dates align between the two calendars, January 1st is the same day for both calendars.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+
+    /**
+     * There are 4 weeks of 7 days, or 28 total days in a month.
+     */
+    static final int DAYS_IN_MONTH = WEEKS_IN_MONTH * DAYS_IN_WEEK;
+
+    /**
+     * There are 13 months of 28 days, or 364 days in a (non-leap) year.
+     */
+    static final int DAYS_IN_YEAR = MONTHS_IN_YEAR * DAYS_IN_MONTH;
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: There are 365 days in the year, as every year has a "year day" (which falls on December 31st for the ISO calendar).

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+
+    /**
+     * Range of proleptic month.
+     */
+    private static final ValueRange PROLEPTIC_MONTH_RANGE = ValueRange.of (0, 1_000_000 * 13L - 1);
+
+    /**
+     * Range of day of month.
+     */
+    static final ValueRange DAY_OF_MONTH_RANGE = ValueRange.of (1, DAYS_IN_MONTH, DAYS_IN_MONTH + 1);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: You probably need to use 0/29, to allow for the year-day and leap-day.  And something similar for month-of-year.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+    /**
+     * Range of day of year.
+     */
+    static final ValueRange DAY_OF_YEAR_RANGE = ValueRange.of (1, DAYS_IN_YEAR, DAYS_IN_YEAR + 1);
+
+    /**
+     * Range of month of year.
+     */
+    static final ValueRange MONTH_OF_YEAR_RANGE = ValueRange.of (1, MONTHS_IN_YEAR);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedChronology.java]: The "extra" days (year-day and leap-day) aren't part of any month, and need their own.  They are also **not** part of any week

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 14]
```diff
@@ -0,0 +1,624 @@
...
+     * The month.
+     */
+    private final short month;
+
+    /**
+     * The day.
+     */
+    private final short day;
+
+    /**
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: Whether or not a day is the leap day or year day should be trivially derivable from month/day values, and probably doesn't need to be stored.  If you do, mark them `transient`, as you need to recalc this after deserialization anyways.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 15]
```diff
@@ -0,0 +1,624 @@
...
+     * Creates an instance from validated data.
+     *
+     * @param prolepticYear the International fixed proleptic-year
+     * @param month         the International fixed month, from 1 to 13
+     * @param dayOfMonth    the International fixed day-of-month, from 1 to 28, the 29th is only legal for leap day and year day
+     */
+    private InternationalFixedDate (final int prolepticYear, final int month, final int dayOfMonth) {
+        ChronoField.YEAR.checkValidValue (prolepticYear);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: `ChronoField.YEAR` is going to allow negative values.  You created your own `YEAR_RANGE`, right?  I personally am leery of including anything with a decent chance of throwing an exception in a constructor, because you can get into weird places if the caller catches it (the reference might be returned, which means you may end up with a sorta-useable object...); this is part of what the `.create(...)` method was for.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 16]
```diff
@@ -0,0 +1,624 @@
...
+        }
+
+        this.prolepticYear = prolepticYear;
+        this.month = (short) month;
+        this.day = (short) dayOfMonth;
+        this.isLeapDay = dayOfMonth == 29 && month == 6;
+        this.isYearDay = dayOfMonth == 29 && month == 13;
+
+        if (this.isLeapDay && !isLeapYear ()) {
+            throw new DateTimeException ("Invalid leap date: " + prolepticYear + '-' + month + '-' + dayOfMonth);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: This message should probably state the given year isn't actually a leap year.  It would be better to pass `prolepticYear` to the static method in the chronology, too.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 17]
```diff
@@ -0,0 +1,624 @@
...
+package org.threeten.extra.chrono;
+
+import java.io.Serializable;
+import java.time.*;
+import java.time.chrono.ChronoLocalDate;
+import java.time.chrono.ChronoLocalDateTime;
+import java.time.chrono.ChronoPeriod;
+import java.time.temporal.*;
+
+/**
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: You need to document the month/day-of-month values that designate the year-day and leap-day.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 18]
```diff
@@ -0,0 +1,624 @@
...
+     * @return the date in International fixed calendar system, not null
+     * @throws DateTimeException if the value of any field is out of range,
+     *                           or if the day-of-year is invalid for the year
+     */
+    static InternationalFixedDate ofYearDay (final int prolepticYear, final int dayOfYear) {
+        ChronoField.YEAR.checkValidValue (prolepticYear);
+        InternationalFixedChronology.DAY_OF_YEAR_RANGE.checkValidValue (dayOfYear, ChronoField.DAY_OF_YEAR);
+        boolean leap = InternationalFixedChronology.INSTANCE.isLeapYear (prolepticYear);
+
+        if (dayOfYear > (InternationalFixedChronology.DAYS_IN_YEAR + 1) && !leap) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: You can use `=` here, as you already check for values greater than.  (If `DAYS_IN_YEAR` actually contained the correct value, this particular check would return the wrong results).

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 19]
```diff
@@ -0,0 +1,624 @@
...
+     */
+    public static InternationalFixedDate ofEpochDay (final long epochDay) {
+        ChronoField.EPOCH_DAY.range ().checkValidValue (epochDay, ChronoField.EPOCH_DAY);
+        long zeroDay = epochDay + InternationalFixedChronology.DAYS_0000_TO_1970;
+
+        if (zeroDay < 0) {
+            throw new DateTimeException ("Negative epoch invalid: " + epochDay);
+        }
+
+        long yearEst = (400 * zeroDay) / DAYS_PER_CYCLE;
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: This is labelled as an estimate, but I see no attempt (or comment about the lack) to true it up.  My gut is uneasy about the calculation I'm seeing, although I don't have anything to back that up.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 20]
```diff
@@ -0,0 +1,624 @@
...
+     * @param epochDay the epoch day to convert based on 1970-01-01 (ISO)
+     * @return the date in International fixed calendar system, not null
+     * @throws DateTimeException if the epoch-day is out of range
+     */
+    public static InternationalFixedDate ofEpochDay (final long epochDay) {
+        ChronoField.EPOCH_DAY.range ().checkValidValue (epochDay, ChronoField.EPOCH_DAY);
+        long zeroDay = epochDay + InternationalFixedChronology.DAYS_0000_TO_1970;
+
+        if (zeroDay < 0) {
+            throw new DateTimeException ("Negative epoch invalid: " + epochDay);
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: Presumably, you'll get negative _epoch_ days, so this message just means confusion for your end user.  You should just let the later year-validation check handle this case, anyways.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 21]
```diff
@@ -0,0 +1,624 @@
...
+
+        // check year now we are certain it is correct
+        int year = ChronoField.YEAR.checkValidIntValue (yearEst);
+
+        return new InternationalFixedDate (year, month, dom);
+    }
+
+    private static InternationalFixedDate resolvePreviousValid (final int prolepticYear, final int month, final int day) {
+        int monthR = Math.min (month, InternationalFixedChronology.MONTHS_IN_YEAR);
+        int dayR = Math.min (day, InternationalFixedChronology.DAYS_IN_MONTH + (month == 6 || month == 13 ? 1 : 0));
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: The max value for day-of-month isn't taking into account if the year is actually a leap-year.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 22]
```diff
@@ -0,0 +1,624 @@
...
+        return month;
+    }
+
+    @Override
+    int getDayOfMonth () {
+        return day;
+    }
+
+    @Override
+    int getDayOfYear () {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: Returns the wrong result after a leap-day.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 23]
```diff
@@ -0,0 +1,624 @@
...
+        return ValueRange.of (1, InternationalFixedChronology.WEEKS_IN_MONTH);
+    }
+
+    @Override
+    InternationalFixedDate resolvePrevious (final int newYear, final int newMonth, final int dayOfMonth) {
+        return resolvePreviousValid (newYear, newMonth, dayOfMonth);
+    }
+
+    @Override
+    public ValueRange range (final TemporalField field) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: You're going to need special values for day-of-week (and a few others), given that the leap-day and year-day aren't part of the week.  See how I handled this in the Discordian calendar (and some of the revisions Joda-Stephen made for related methods).

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 24]
```diff
@@ -0,0 +1,624 @@
...
+        return (ChronoLocalDateTime<InternationalFixedDate>) ChronoLocalDate.super.atTime (localTime);
+    }
+
+    @Override
+    public long until (final Temporal endExclusive, final TemporalUnit unit) {
+        return until (InternationalFixedDate.from (endExclusive), unit);
+    }
+
+    @Override
+    long until (final AbstractDate end, final TemporalUnit unit) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: `WEEKS` is a pain, given that the leap-day and year-day aren't part of the regular week.  Again, look at how the Discordian Chronology deals with this.

File: src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java [hunk 25]
```diff
@@ -0,0 +1,624 @@
...
+        return super.until (end, unit);
+    }
+
+    /**
+     * Get the number of years from this date to the given day.
+     *
+     * @param end The end date.
+     * @return The number of years from this date to the given day.
+     */
+    long yearsUntil (final InternationalFixedDate end) {
```
Comment (Clockwork-Muse) [inline src/main/java/org/threeten/extra/chrono/InternationalFixedDate.java]: Unfortunately, the calculation is more complicated than this - you _aren't_ "simulating the effect of inserting the leap day".

File: src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java [hunk 26]
```diff
@@ -0,0 +1,794 @@
...
+        assertEquals (InternationalFixedChronology.INSTANCE.range (ChronoField.DAY_OF_YEAR), ValueRange.of (1, 364, 365));
+        assertEquals (InternationalFixedChronology.INSTANCE.range (ChronoField.MONTH_OF_YEAR), ValueRange.of (1, 13));
+    }
+
+    //-----------------------------------------------------------------------
+    // InternationalFixedDate.range
+    //-----------------------------------------------------------------------
+    @DataProvider (name = "ranges")
+    Object[][] data_ranges () {
+        return new Object[][] {
```
Comment (Clockwork-Muse) [inline src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java]: You don't have all range tests needed for dealing with the added, non-week/non-month days.

File: src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java [hunk 27]
```diff
@@ -0,0 +1,794 @@
...
+    public void test_getLong_unsupported () {
+        InternationalFixedDate.of (2012, 6, 28).getLong (ChronoField.MINUTE_OF_DAY);
+    }
+
+    //-----------------------------------------------------------------------
+    // InternationalFixedDate.with
+    //-----------------------------------------------------------------------
+    @DataProvider (name = "with")
+    Object[][] data_with () {
+        return new Object[][] {
```
Comment (Clockwork-Muse) [inline src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java]: No tests dealing with year-day/leap-day.

File: src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java [hunk 28]
```diff
@@ -0,0 +1,794 @@
...
+        LocalDateTime test = LocalDateTime.MIN.with (date);
+        assertEquals (test, LocalDateTime.of (2012, 6, 4, 0, 0));
+    }
+
+    //-----------------------------------------------------------------------
+    // InternationalFixedDate.plus
+    //-----------------------------------------------------------------------
+    @DataProvider (name = "plus")
+    Object[][] data_plus () {
+        return new Object[][] {
```
Comment (Clockwork-Muse) [inline src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java]: What happens when you add a week to the leap-day/year-day?  Remember, they're not part of the normal week (or month, for that matter) - Discordian goes into this some, too.

File: src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java [hunk 29]
```diff
@@ -0,0 +1,794 @@
...
+
+                { 2012, 13, 6, 3, ChronoUnit.MONTHS, 2013, 3, 6 },
+                { 2011, 13, 26, 1, ChronoUnit.YEARS, 2012, 13, 26 },
+                { 2014, 13, 26, -2, ChronoUnit.YEARS, 2012, 13, 26 },
+                { 2012, 13, 26, -6, ChronoUnit.YEARS, 2006, 13, 26 },
+                { 2012, 13, 6, -6, ChronoUnit.YEARS, 2006, 13, 6 },
+        };
+    }
+
+    @DataProvider (name = "plusLeap")
```
Comment (Clockwork-Muse) [inline src/test/java/org/threeten/extra/chrono/TestInternationalFixedChronology.java]: You may as well test the year-day here, too (I originally did these separate data constructors because dealing with leap stuff isn't bi-directional).  Except you're not testing leap-day stuff anyways...

**Comment by Clockwork-Muse** [comment]
Some of the files are in a different order than the existing implementations (there's a consistent grouping), which you'll also need to change.

**Comment by catull** [comment]
I created a branch to address all the review findings.
Would like to close this request and create a new PR.
